File: internal/pkg/pinyin/lex.go
package pinyin

import (
	"github.com/go-errors/errors"
)

func Lex(raw string) ([][]Token, error) {
	result := [][]Token{}
	tokens := []Token{}
	for raw != "" {
		groups := pinyinRE.FindStringSubmatch(raw)
		if groups == nil {
			return nil, errors.Errorf("%q: invalid pinyin", raw)
		}
		switch groups[1] {
		case "":
			tokens = append(tokens, Token{
				syllable: groups[2],
				tones:    newTonesFromString(groups[3]),
			})
		case "/":
			result = append(result, tokens)
			tokens = []Token{}
		default:
			tokens = append(tokens, Token{syllable: groups[1]})
		}
		raw = raw[len(groups[0]):]
	}
	result = append(result, tokens)
	return result, nil
}

File: electron2/src/pinyin/Lex.ts
import Token from './Token';
import Tokens from './Tokens';
import Tones from './Tones';
import PinyinRE from './PinyinRE';

export default function Lex(raw: string): Tokens[] {
	const ret: Tokens[] = [];
	let tokens: Token[] = [];
	while (raw) {
		const groups = raw.match(PinyinRE);
		if (!groups) {
			throw new Error(`${raw}: invalid pinyin`)
		}
		switch (groups[1]) {
		case undefined:
			tokens.push(new Token(groups[2], new Tones(groups[3])));
			break;
		case "/":
			ret.push(new Tokens(tokens));
			tokens = [];
			break;
		default:
			tokens.push(new Token(groups[1]))
		}
		raw = raw.slice(groups[0].length);
	}
	ret.push(new Tokens(tokens));
	return ret
}
